File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = 0
        max_length = 0
        char_index = {}  # Map characters to their last indices

        for end in range(len(s)):
            if s[end] in char_index and char_index[s[end]] >= start:
                start = char_index[s[end]] + 1
            char_index[s[end]] = end
            max_length = max(max_length, end - start + 1)

        return max_length
solution=Solution()
s="abcabcbb"
res=solution.lengthOfLongestSubstring(s)
             

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let start = 0;
    let maxLength = 0;
    let charIndex = new Map(); // Stores the last seen index of characters

    for (let end = 0; end < s.length; end++) {
        const currentChar = s[end];
        if (charIndex.has(currentChar) && charIndex.get(currentChar) >= start) {
            start = charIndex.get(currentChar) + 1;
        }
        charIndex.set(currentChar, end);
        maxLength = Math.max(maxLength, end - start + 1);
    }

    return maxLength;
};